### ISSUE TITLE
[BUG] HA mode is not enabled after setting up a Harvester cluster with 3 nodes

### ISSUE THREAD

**Issue by yas51ne** [harvester/harvester#6546]
**Describe the bug**
After setting up a Harvester cluster with 3 nodes (harvester-01, harvester-02, and harvester-03), I encountered issues where new nodes successfully join the cluster but are not recognized as control-plane or etcd nodes. As a result, I am unable to place the first node (harvester-01) into maintenance mode because the cluster seems to still be operating in a single control-plane configuration, despite the additional nodes.

**To Reproduce**
Set up a Harvester cluster with 3 nodes:

- harvester-01: Control-plane, etcd, master
- harvester-02: Worker (attempting to add as control-plane and etcd)
- harvester-03: Worker (also attempting to add as control-plane and etcd)

The new nodes (harvester-02 and harvester-03) join the cluster but are not promoted to control-plane or etcd roles.
Attempting to manually add the roles using kubectl label does not schedule etcd or control-plane pods on the new nodes.

- Create config:
```yaml
scheme_version: 1
token: ....
os:
  hostname: harvester-01
  ssh_authorized_keys:
    - github:...
    - "ssh-rsa ..."
  password: .....
  dns_nameservers:
    - 192.168.100.1
    - 8.8.8.8
    - 1.1.1.1
  ntp_servers:
  - 192.168.100.1
  - 0.suse.pool.ntp.org
  - 1.suse.pool.ntp.org
  labels:
    topology.kubernetes.io/zone: zone1
    region: europecentral
    cloud: homelab.ai
install:
  mode: create
  role: default
  automatic: true
  skipchecks: true
  management_interface:
    interfaces:
    - name: enp4s0
    default_route: true
    method: static
    ip: 192.168.100.10
    subnet_mask: 255.255.255.0
    gateway: 192.168.100.1
    bond_options:
      mode: balance-tlb
      miimon: 100
  device: /dev/nvme0n1       # The target disk to install
  iso_url: http://192.168.200.138:8080/asset-mirror/releases/download/http/harvester/harvester-v1.3.2-amd64.iso
  #tty: tty1 #,115200n8  # For machines without a VGA console

  vip: 192.168.100.9
  vip_mode: static
  vip_hw_addr: ""

```
- Join config:
```yaml
scheme_version: 1
server_url: https://192.168.100.9:443
token: .....
os:
  hostname: harvester-03
  ssh_authorized_keys:
    - github:...
    - "ssh-rsa ..."
  password: .... 
  dns_nameservers:
    - 192.168.100.1
    - 8.8.8.8
    - 1.1.1.1
  ntp_servers:
  - 192.168.100.1
  - 0.suse.pool.ntp.org
  - 1.suse.pool.ntp.org
  labels:
    topology.kubernetes.io/zone: zone1
    region: europecentral
    cloud: homelab.ai
install:
  mode: join
  automatic: true
  skipchecks: true
  management_interface:
    interfaces:
    - name: enp4s0
    default_route: true
    method: static
    ip: 192.168.100.12
    subnet_mask: 255.255.255.0
    gateway: 192.168.100.1
    bond_options:
      mode: balance-tlb
      miimon: 100
  device: /dev/nvme0n1       # The target disk to install
  iso_url: http://192.168.200.138:8080/asset-mirror/releases/download/http/harvester/harvester-v1.3.2-amd64.iso
  tty: ttyS1,115200n8   # For machines without a VGA console
```
**Expected behavior**
The additional nodes should be promoted to control-plane and etcd roles, resulting in a High Availability (HA) cluster where all control-plane nodes can be put into maintenance mode independently.

**Support bundle**
[supportbundle_282cf640-adc6-47d4-aed4-766a4cd7d751_2024-09-11T19-30-38Z.zip](https://github.com/user-attachments/files/16969247/supportbundle_282cf640-adc6-47d4-aed4-766a4cd7d751_2024-09-11T19-30-38Z.zip)

**Environment**
 - Harvester ISO version: 1.3.2
 - Underlying Infrastructure: Baremetal with ACEMAGICIAN AM06 Pro

**Additional context**
Actual Behavior

- The additional nodes (harvester-02 and harvester-03) join the cluster but are only recognized as worker nodes.
- The first node (harvester-01) remains the sole control-plane and etcd node.
- When attempting to place harvester-01 into maintenance mode, I receive the error: Single control-plane cluster, cannot place control-plane in maintenance mode.
- No etcd pods are scheduled on harvester-02 or harvester-03 despite manually applying control-plane and etcd labels.

Screenshots

- Harvester UI - Hosts Overview:
harvester-02 and harvester-03 appear as active but do not have control-plane or etcd labels.
<img width="1505" alt="Screenshot 2024-09-11 at 21 45 31" src="https://github.com/user-attachments/assets/c2bdea1b-4523-46f5-963a-f92ffb9d882e">

- Rancher UI - Nodes Overview:
The nodes are visible, but only harvester-01 has the control-plane and etcd roles.
<img width="1500" alt="Screenshot 2024-09-11 at 21 45 16" src="https://github.com/user-attachments/assets/d4945ef4-8fe1-4474-9e8b-047be2de6a3c">

- Logs and Outputs

kubectl get nodes : Output showing that the additional nodes are labeled as control-plane and etcd but no etcd or control-plane pods are scheduled.
```bash
harvester-01:/home/rancher # kubectl get nodes
NAME           STATUS   ROLES                       AGE     VERSION
harvester-01   Ready    control-plane,etcd,master   5h47m   v1.28.12+rke2r1
harvester-02   Ready    <none>                      48m     v1.28.12+rke2r1
harvester-03   Ready    <none>                      4h3m    v1.28.12+rke2r1
```

kubectl get pods -n kube-system -l component=etcd -o wide: Output showing etcd pods running only on harvester-01.
```bash
harvester-01:/home/rancher # kubectl get pods -n kube-system -l component=etcd -o wide
NAME                READY   STATUS    RESTARTS   AGE   IP               NODE           NOMINATED NODE   READINESS GATES
etcd-harvester-01   1/1     Running   2          80m   192.168.100.10   harvester-01   <none>           <none>
```

etcdctl member list: Output showing only harvester-01 as a member of the etcd cluster.
```bash
harvester-01:/home/rancher # kubectl exec -n kube-system etcd-harvester-01 -- etcdctl   --endpoints=https://127.0.0.1:2379   --cacert=/var/lib/rancher/rke2/server/tls/etcd/server-ca.crt   --cert=/var/lib/rancher/rke2/server/tls/etcd/server-client.crt   --key=/var/lib/rancher/rke2/server/tls/etcd/server-client.key  member list
c2bde6fe848e75ba, started, harvester-01-933e1269, https://192.168.100.10:2380, https://192.168.100.10:2379, false
```

I have attempted the following steps:

- Recreating nodes (harvester-02 and harvester-03) and ensuring they join the cluster.
- Manually adding control-plane and etcd roles via kubectl label.

Despite these efforts, the nodes remain as worker nodes, and etcd/control-plane pods are not scheduled on them.

**Comment by bk201** [comment]
@FrankYang0529 Please help check! Thanks.

**Comment by FrankYang0529** [comment]
I think the root cause is network issue on node harvester-03. We can see harvester pod log on node harvester-03:

```
E0911 19:28:19.748591       7 reflector.go:140] github.com/rancher/steve/pkg/clustercache/controller.go:165: Failed to watch *summary.SummarizedObject: failed to list *summary.Summarize
dObject: Get "https://10.53.0.1:443/apis/crd.projectcalico.org/v1/clusterinformations?resourceVersion=243220": dial tcp 10.53.0.1:443: connect: connection refused
W0911 19:28:20.001800       7 reflector.go:424] github.com/rancher/steve/pkg/clustercache/controller.go:165: failed to list *summary.SummarizedObject: Get "https://10.53.0.1:443/apis/no
de.harvesterhci.io/v1beta1/ksmtuneds?resourceVersion=243422": dial tcp 10.53.0.1:443: connect: connection refused
E0911 19:28:20.001873       7 reflector.go:140] github.com/rancher/steve/pkg/clustercache/controller.go:165: Failed to watch *summary.SummarizedObject: failed to list *summary.Summarize
dObject: Get "https://10.53.0.1:443/apis/node.harvesterhci.io/v1beta1/ksmtuneds?resourceVersion=243422": dial tcp 10.53.0.1:443: connect: connection refused
W0911 19:28:20.057018       7 reflector.go:424] github.com/rancher/steve/pkg/clustercache/controller.go:165: failed to list *summary.SummarizedObject: Get "https://10.53.0.1:443/apis/ma
nagement.cattle.io/v3/nodedrivers?resourceVersion=243385": dial tcp 10.53.0.1:443: connect: connection refused
```

Pod kube-proxy-harvester-03 also has log:
```
E0911 18:50:44.339546       1 reflector.go:147] k8s.io/client-go/informers/factory.go:150: Failed to watch *v1.Service: failed to list *v1.Service: Get "https://127.0.0.1:6443/api/v1/se
rvices?labelSelector=%21service.kubernetes.io%2Fheadless%2C%21service.kubernetes.io%2Fservice-proxy-name&resourceVersion=217866": net/http: TLS handshake timeout - error from a previous
 attempt: read tcp 127.0.0.1:35942->127.0.0.1:6443: read: connection reset by peer
I0911 18:51:34.818837       1 trace.go:236] Trace[676891285]: "Reflector ListAndWatch" name:k8s.io/client-go/informers/factory.go:150 (11-Sep-2024 18:51:18.424) (total time: 16394ms):
Trace[676891285]: ---"Objects listed" error:<nil> 16394ms (18:51:34.818)
Trace[676891285]: [16.394287733s] [16.394287733s] END
I0911 18:51:34.820092       1 trace.go:236] Trace[1386573943]: "Reflector ListAndWatch" name:k8s.io/client-go/informers/factory.go:150 (11-Sep-2024 18:51:08.125) (total time: 26694ms):
Trace[1386573943]: ---"Objects listed" error:<nil> 26694ms (18:51:34.820)
Trace[1386573943]: [26.694733803s] [26.694733803s] END
E0911 18:52:25.801303       1 reflector.go:147] k8s.io/client-go/informers/factory.go:150: Failed to watch *v1.Service: unknown (get services) - error from a previous attempt: read tcp
127.0.0.1:36474->127.0.0.1:6443: read: connection reset by peer
E0911 18:52:25.801328       1 reflector.go:147] k8s.io/client-go/informers/factory.go:150: Failed to watch *v1.EndpointSlice: unknown (get endpointslices.discovery.k8s.io) - error from
a previous attempt: read tcp 127.0.0.1:36372->127.0.0.1:6443: read: connection reset by peer
E0911 18:52:25.801428       1 reflector.go:147] k8s.io/client-go/informers/factory.go:150: Failed to watch *v1.Node: unknown (get nodes) - error from a previous attempt: read tcp 127.0.
0.1:36342->127.0.0.1:6443: read: connection reset by peer
```

**Comment by yas51ne** [comment]
communication between nodes is whitelisted. and no filtering is applied at the node level. otherwise the node would never join the cluster.
and the connectivity issue i see in the logs is when it is trying to connect to localhost : >127.0.0.1:6443: read: connection reset by peer.
reste by peer indicated more a response from the server not establishing the connection than network issue, maybe certificate?

**Comment by ibrokethecloud** [comment]
```apiVersion: v1
items:
- apiVersion: v1
  kind: Endpoints
  metadata:
    creationTimestamp: "2024-09-11T13:54:15Z"
    labels:
      endpointslice.kubernetes.io/skip-mirror: "true"
    managedFields:
    - apiVersion: v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:labels:
            .: {}
            f:endpointslice.kubernetes.io/skip-mirror: {}
        f:subsets: {}
      manager: kube-apiserver
      operation: Update
      time: "2024-09-11T19:28:23Z"
    name: kubernetes
    namespace: default
    resourceVersion: "243953"
    uid: 9e6a819e-18b5-48c4-ab61-99a46576fe2b
  subsets:
  - addresses:
    - ip: 192.168.100.10
    ports:
    - name: https
      port: 6443
      protocol: TCP
kind: List
metadata:
  resourceVersion: "247116"
```

I suspect nodes are unable to talk to harvester-01 node on 6443. 

Are you able to run a curl from harvester-03 to node harvester-01 on port 6443?

**Comment by yas51ne** [comment]
yes this works otherwise kubelet will not be able to report back to kubapi:

```bash
rancher@harvester-03:~> curl https://192.168.100.10:6443 -k
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
}
rancher@harvester-03:~> logout
Connection to 192.168.100.12 closed.

rancher@harvester-02:~> curl https://192.168.100.10:6443 -k
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
}
rancher@harvester-02:~>
``` 

Can  topology.kubernetes.io/zone have an impact here? the 3 nodes has the same  topology.kubernetes.io/zone value so they will be considered in the same zone.

**Comment by yas51ne** [comment]
is it possible to update the node role manually to test if it will change?

**Comment by yas51ne** [comment]
After further investigation, I identified the root cause of the issue where harvester-02 and harvester-03 were not joining as control-plane and etcd nodes.

**Root Cause:**
The issue was related to the Kubernetes topology zone labels on each node. All nodes were set to the same zone label (topology.kubernetes.io/zone), which seemed to cause conflicts and prevent the new nodes from properly joining as management nodes.

**Solution:**
To resolve the issue, I updated the topology zone labels for harvester-02 and harvester-03 to have unique values, making sure that each node had a distinct zone label. After applying the changes, the nodes successfully joined the cluster as control-plane and etcd nodes, and everything is now functioning as expected.

**Steps Taken:**
Verified the topology zone labels for each node using:
```bash
kubectl get nodes --show-labels
```
Updated the zone labels on harvester-02 and harvester-03 to ensure they were unique:
```bash
kubectl label node harvester-02 topology.kubernetes.io/zone=zone2 --overwrite
kubectl label node harvester-03 topology.kubernetes.io/zone=zone3 --overwrite
```
After making these changes, the nodes were successfully promoted to control-plane and etcd roles.
